README.md
# Amok(1)
[![tips](https://img.shields.io/gratipay/caspervonb.svg?style=flat-square)](https://gratipay.com/caspervonb/)
[![chat](https://img.shields.io/badge/gitter-join%20chat-green.svg?style=flat-square)](https://gitter.im/caspervonb/amok)
[![npm](https://img.shields.io/npm/v/amok.svg?style=flat-square)](https://www.npmjs.org/package/amok)

## Synopsis
```sh
amok [options] <script>
```

## Description
Amok standalone command line tool for rapid prototyping and development of JavaScript applications.

It monitors changes in the file system. As soon as you save a file, it is then preprocessed, compiled and bundled as needed, and re-compiled in the client session without refreshing or restarting the client.

This re-compilation is done through a debugging session, unlike reloading or reevaluation, re-compilation leaves the application state intact, no side effects are executed when doing re-compilation.

Additional features include a zero configuration http development server for developing front end applications, an interactive mode (read–eval–print loop).

## Example
### Setting Chrome as the Client
First, set `AMOK_CLIENT` to the path of the chrome executable, the
appropriate value depends on your operating system.

This could also be set by passing it via the `--client` option.

#### Linux
```sh
export AMOK_CLIENT='google-chrome --remote-debugging-port=9222'
```

#### OSX
```sh
export AMOK_CLIENT='"/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome" --remote-debugging-port=9222'
```

#### Windows 7 (32 bit)
```sh
SET AMOK_CLIENT='"C:\Program Files\Google\Application\chrome.exe" --remote-debugging-port=9222'
```
#### Windows 7 (64 bit)
```sh
SET AMOK_CLIENT='"C:\Program Files\Google\Application\chrome.exe" --remote-debugging-port=9222'
```

### Launching with a Compiler
To lanch with a compiler, we'll set the --compiler option (default can be set via the `AMOK_COMPILER` environment variable)

### Browserify
```sh
amok --compiler 'watchify -o $@' entry.js
```

### Webpack
```sh
amok --compiler 'webpack --watch --output-file $@' entry.js
```

## Options
```
-h, --host <HOST>
  specify the http host, default HOST is localhost.

-p, --port <PORT>
  specify the http port, default PORT is 9966.

-H, --debugger-host <HOST>
  specify the remote debugger host, default HOST is localhost.

-P, --debugger-port <PORT>
  specify the remote debugger port, default PORT is 9222.

--client
  specify the client to spawn

--compiler
  specify the compiler to spawn

-v, --verbose
  enable verbose logging mode

--no-client
  disable client

--no-compiler
  disable compiler
```

Amok requires that a client is listening on the remote debugging port when launching, it can spawn a client for you at the appropriate time, this is set by passing the `--client` option with the executable name and appropriate flags, this option has automatic variables available to it.

Amok can also, optionally use a compiler to process script sources, this compiler is specified via the `--compiler` option, this option has automatic variables available to it.

Any extra arguments following the `--` terminator, will be passed as arguments when spawning the compiler, if one is specified.

## Environment Variables
These environment variables are used to provide amok with default values.

```
AMOK_CLIENT
  When set to a executable, will be used as the default client value.

AMOK_COMPILER
  When set to an executable, will be used as the default compiler value.
```

## Automatic Variables
These automatic variables are set and substituted when spawning clients and compilers.

```
$@
  When using a compiler, this is set to the output path of the compilation result
```


changelog.md
# CHANGELOG

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


cmd.js
#!/usr/bin/env node

var amok = require('./');
var async = require('async');
var cmd = require('commander');
var fs = require('fs');
var path = require('path');
var temp = require('temp');
var repl = require('repl');
var util = require('util');
var colors = require('colors');

var pkg = require('./package.json');

cmd.usage('[options] <script>');
cmd.version(pkg.version);

cmd.option('--host <HOST>', 'specify http host', 'localhost');
cmd.option('--port <PORT>', 'specify http port', 9966);

cmd.option('--debugger-host <HOST>', 'specify debugger host', 'localhost');
cmd.option('--debugger-port <PORT>', 'specify debugger port', 9222);

cmd.option('-i, --interactive', 'enable interactive mode');

cmd.option('--client <CMD>', 'specify the client to spawn');
cmd.option('--compiler <CMD>', 'specify the compiler to spawn');

cmd.option('-v, --verbose', 'enable verbose logging mode');

cmd.parse(process.argv);

cmd.cwd = process.cwd();

function compiler(callback, data) {
  if (cmd.compiler === undefined) {
    cmd.scripts = cmd.args.reduce(function(object, value, key) {
      object[value] = path.resolve(value);
      return object;
    }, {});

    return callback(null, null);
  }

  if (cmd.verbose) {
    console.info('Spawning compiler...');
  }

  var compiler = amok.compile(cmd, function(error, stdout, stderr) {
    if (error) {
      return callback(error);
    }

    stdout.pipe(process.stdout);
    stderr.pipe(process.stderr);

    cmd.scripts = {
      'bundle.js': compiler.output,
    };

    callback(null, compiler);
  });
}

function watcher(callback, data) {
  for (file in cmd.scripts) {
    fs.lstat(cmd.scripts[file], function(err, stats) {
      if (err && err.code === 'ENOENT') {
        console.error('[ERROR] Input file "%s" was not found'.red, file);
      }
    });
  }

  var watcher = amok.watch(cmd, function() {
    if (cmd.verbose) {
      console.info('File watcher ready');
    }

    watcher.on('change', function(filename) {
      if (cmd.verbose) {
        console.info(filename, 'changed');
      }

      var script = Object.keys(cmd.scripts)
        .filter(function(key) {
          return cmd.scripts[key] === filename
        })[0];

      if (script) {
        if (cmd.verbose) {
          console.info('Re-compiling', script, '...');
        }

        fs.readFile(filename, 'utf-8', function(error, contents) {
          if (error) {
            return console.error(error);
          }

          data.bugger.source(script, contents, function(error) {
            if (error) {
              return console.error(error);
            }

            if (cmd.verbose) {
              console.info('Re-compilation succesful');
            }
          });
        });
      }
    });

    callback(null, watcher);
  });
}

function server(callback, data) {
  if (cmd.verbose) {
    console.info('Starting server...');
  }

  var server = amok.serve(cmd, function() {
    var address = server.address();
    if (cmd.verbose) {
      console.info('Server listening on http://%s:%d', address.address,
        address.port);
    }

    callback(null, server);
  });
}

function client(callback, data) {
  if (cmd.client === undefined) {
    return callback(null, null);
  }

  if (cmd.verbose) {
    console.log('Spawning client...');
  }

  var client = amok.open(cmd, function(error, stdout, stderr) {
    if (error) {
      return callback(error);
    }

    stdout.pipe(process.stdout);
    stderr.pipe(process.stderr);

    callback(null, client);
  });
}

function bugger(callback, data) {
  if (cmd.verbose) {
    console.info('Attaching debugger...');
  }

  var bugger = amok.debug(cmd, function(error, target) {
    if (error) {
      return callback(error);
    }

    if (cmd.verbose) {
      console.info('Debugger attached to', target.title);
    }

    bugger.console.on('data', function(message) {
      if (message.parameters) {
        var parameters = message.parameters.map(function(parameter) {
          return parameter.value;
        });

        if (console[message.level]) {
          console[message.level].apply(console, parameters);
        }
      }
    });

    callback(null, bugger);
  });
}

function prompt(callback, data) {
  if (cmd.interactive === undefined) {
    return callback(null, null);
  }

  var options = {
    prompt: '> ',
    input: process.stdin,
    output: process.stdout,
    writer: function(result) {
      if (result.value) {
        return util.inspect(result.value, {
          colors: true,
        });
      } else if (result.objectId) {
        return util.format('[class %s]\n%s', result.className, result.description);
      }
    },

    eval: function(cmd, context, filename, write) {
      data.bugger.evaluate(cmd, function(error, result) {
        write(error, result);
      });
    },
  };

  var prompt = repl.start(options);
  callback(null, prompt);
}

async.auto({
  'compiler': [compiler],
  'server': ['compiler', server],
  'client': ['server', client],
  'bugger': ['client', bugger],
  'watcher': ['bugger', watcher],
  'prompt': ['bugger', prompt],
}, function(error, data) {
  if (error) {
    console.error(error);
    process.exit(1);
  }
});
